### ISSUE TITLE
[RFC] FastCall API

### ISSUE THREAD

**Issue by sjinks** [zephir-lang/zephir#1510]
Currently, Zephir compiler misses a lot of cases where method caching could save some time. Some of these misses are by design (i.e., when a non-final method is called, the compiler cannot be sure which class `$this` refers to), some are compiler defects (when the same method is called in a loop).

Some of these deficiencies are dealt with by the Zephir Kernel (it tries to cache functions and methods called by the user code). However, this caching is not very efficient: a) every method/function is cached, b) cache is stored in hash tables.

FastCall API can be used to deal with these deficiencies; however, *right now* it is quite low-level and is for manual use only.

[Usage example](https://github.com/sjinks/zephir/commit/c360d861dcb2585431ae53fe9483e1f87f6600cb#diff-b913b909037a897c48b0d7f9e284dcf2R141)

[Benchmarks](https://travis-ci.org/sjinks/zephir/jobs/222133655#L4901)

PS: `fastcall` branch was branched off `master` and does not include fixes from #1509.

However, my initial impression is that it could be faster than our current CALL API.

The main question is: do we need it? :-)

@sergeyklay @Jurigag @ovr

**Comment by sjinks** [comment]
The current implementation is very simple:

```c
void fcall_init(zval *return_value, zval *callable)
{
	zephir_fci *info = ecalloc(1, sizeof(zephir_fci));
	char *error = NULL;

	if (FAILURE == zend_fcall_info_init(callable, IS_CALLABLE_STRICT, &info->fci, &info->fcc, NULL, &error)) {
		zend_throw_exception_ex(zend_exception_get_default(), 0, "Error: %s", error);
		efree(error);
		RETURN_NULL();
	}

	ZVAL_PTR(return_value, info);
}

void fcall_call(zval *return_value, zval *fcall, zval *parameters)
{
	if (parameters && Z_TYPE_P(parameters) != IS_ARRAY && Z_TYPE_P(parameters) != IS_NULL) {
		zend_throw_exception_ex(zend_exception_get_default(), 0, "fcall_call: parameters must be array or NULL");
		RETURN_NULL();
	}

	if (IS_PTR == Z_TYPE_P(fcall)) {
		zephir_fci *info = Z_PTR_P(fcall);

		if (info->fci.size == sizeof(info->fci)) {
			if (parameters && IS_ARRAY == Z_TYPE_P(parameters)) {
				zend_fcall_info_args_ex(&info->fci, info->fcc.initialized ? info->fcc.function_handler : NULL, parameters);
			}

			zend_fcall_info_call(&info->fci, &info->fcc, return_value, NULL);

			if (parameters && IS_ARRAY == Z_TYPE_P(parameters)) {
				zend_fcall_info_args_clear(&info->fci, 1);
			}

			return;
		}
	}

	zend_throw_exception_ex(zend_exception_get_default(), 0, "Bad parameter passed to fcall_call()");
	RETURN_NULL();
}

void fcall_done(zval *fcall)
{
	if (IS_PTR == Z_TYPE_P(fcall)) {
		efree(Z_PTR_P(fcall));
		ZVAL_NULL(fcall);
		return;
	}

	zend_throw_exception_ex(zend_exception_get_default(), 0, "Bad parameter passed to fcall_done()");
}
```

**Comment by sjinks** [comment]
PHP 7.0.17 ZTS DEBUG
```
$this->f()                 5.162
self::f()                  6.259
self::f() -> true          6.289
$this->f() [slow]          3.973
$this->f() [fast]          0.647
self::f() [fast]           0.645
self::f() -> true [fast]   0.643
```

PHP 7.0.17 ZTS DEBUG, #1509:
```
$this->f()         2.074
self::f()          2.210
self::f() -> true  2.222
```

**Comment by sergeyklay** [comment]
Looks good to me. Awesome job man 👍

**Comment by sjinks** [comment]
7.0.17 NTS RELEASE

#1509
```
$this->f()         0.381
self::f()          0.468
self::f() -> true  0.467
```

fastcall
```
$this->f()                 1.187
self::f()                  1.453
self::f() -> true          1.467
$this->f() [slow]          0.650
$this->f() [fast]          0.118
self::f() [fast]           0.117
self::f() -> true [fast]   0.118
```

**Comment by Jurigag** [comment]
`right now it is quite low-level and is for manual use only.`

Will be a problem to make it not manual use? Like automatically used when possible?

**Comment by sjinks** [comment]
Not sure. The reason is that the optimizer does not have enough information about the environment.

For example,

```php
class A
{
    public function someMethod()
    {
        /* ... */
    }

    public function test()
    {
        $this->someMethod();
    }
}
```

The optimizer does not know which `someMethod()` is called: A's or of some other derived class. The developer may know that, and therefore is able to optimize this.

**Comment by sjinks** [comment]
Basically, I see only two safe cases:
1. A plain function is called.
2. A private final method or private static function is called.

**Comment by Jurigag** [comment]
As i understand this is seperated thing from fix here - https://github.com/phalcon/zephir/pull/1509/files ? Or i miss something?

**Comment by sjinks** [comment]
Yes. They are both applicable. #1509 is ready to use, this one is yet to be discussed :-)

**Comment by Jurigag** [comment]
It could be nice to have such a thing. If there are cases where we could make it used by default.

Tbh im not sure what exactly is cached etc stuff, i think like for example in phalcon this would be uselss to add this, i guess the code itself in most of methods is costing more time than calling them.

Though difference on this example of microbench is pretty huge. If you could somehow think a way in compiler to figure out if method call can be fcall optimized or not this could be awesome, because changing code of phalcon will be just huge work, and second code will look worse and not everyone will understand why we use it like this.

**Comment by sjinks** [comment]
@Jurigag it depends

For example, in this case:

```php
foreach ($array as $value) {
    $class->doSomething($value);
}
```

fcall can help. If the same method is going to be used more than once, fcall can help.

Its advantage over our current call API is that it does not waste time maintaining caches and setting up call contexts. The setup is performed once in `fcall_init`. `facll_call` just passes parameters to the method/function and calls its handler, nothing more, this is why it is fast. However, if you pass a fcall variable to another method in another class (it is technically possible), that method will be able to call the cached method even if it is protected or private or is otherwise not callable from the current context.

The disadvantages are that it is too low-level (`fcall_done()` must be called manually for every `fcall_init()` — I failed to make the compiler understand a custom type, it converts it to `zval` :-( ), lacks error checking (by design), and currently have to be used only manually.

**Comment by Jurigag** [comment]
Could it speed up recursive method calling? Like for example this topic on forum - https://forum.phalconphp.com/discussion/15947/acl-increase-speed i have other to solve it, just by adding getters and setters so people can just use some cache to set and remove addRole etc stuff

**Comment by sjinks** [comment]
Yes. You just need to make sure that `fcall_init()` is called only once.

Smth like that:
```php
funciton f($fci = null)
{
    $cleanup = false;
    if (fci === null) {
        $fci = fcall_init('f');
        $cleanup = true;
    }

    /* Recursion */
    $result = fcall_call($fci, [$fci, $other, $parameters]);

    if ($cleanup) {
        fcall_done($fci);
    }
}
```

**Comment by dreamsxin** [comment]
It's more controllable for me.

**Comment by Jurigag** [comment]
Well actually we need to make somehow a nice way to make it default way to call methods, functions etc because currently zephir is around 3 times slower than PHP 7.2 in method calling etc. I would propose somehow to cache all those fcall infos and reuse them, maybe even between requests if somehow possible.

**Comment by Jurigag** [comment]
I think on some point we should focus on this @sergeyklay just test this https://github.com/phalcon/zephir/blob/development/unit-tests/microbench.php between zephir + php 5.6/7.0./7.1/7.2 and implement this in plain php 5.6/7.0/7.1/7.2 you will notice that results for zephir method calling are kind of really meh, because of no fast call. Not sure if im correct but i think it was here on some point? But it was introducing issues, i guess we need to work on this for performance.

**Comment by Jurigag** [comment]
@dreamsxin Could you look into this somehow and check out what could we do here?

**Comment by dreamsxin** [comment]
It is better to directly call the kernel method or the Ç function to reduce the use of zend_call_*

**Comment by Jurigag** [comment]
Yea but it's too complicated to do it in zephir code.

**Comment by sjinks** [comment]
This thing is very complex to implement it properly :-(

When PHP compiles the code, it can calculate FCI fields in advance and make every suitable FCALL reference the respective FCI structure. If we populate the FCI at runtime, there are several challenges:
  * FCI depends on the method visibility (i.e., if we call a protected or private method, we must take into account relationships between classes);
  * we need to take into account magic methods (__invoke, __call, __callStatic), as they allow us to call methods which are not visible to the caller;
  * we store FCIs in a hash; the more elements in that hash, the more expensive cache lookups are (they affect performance; PHP has a pointer to a FCI structure, access time is roughly O(1); we have a hash, access time is roughly O(log n)).

One possible solution is to make Zephir collect all function / method calls in the method being processed, and add the code to populate their FCIs. However,
  * if a method / function is called only sometimes (say, when some condition is true), we will have performance penalty for FCI calculation;
  * FCI won't be shared between methods (say, two methods call `str_replace()` - its FCI will be calculated twice);
  * need to take care of protected / private contexts.

Another solution is to use low-level API, as described above :-(

**Comment by Jurigag** [comment]
Well at least implementing it as low-level API could be fine, to add it in most of the areas where we can achieve actual performance boost, i guess mostly with same method called multiple times.

Though im still wondering how much it would affect performance, like searching all those hashes vs calling it typically.
